File: momentra/momentra/Data/PersonalTransactionRepository.swift
import Foundation

/// GET / PATCH / DELETE loader for personal expense + income transactions.
final class PersonalTransactionRepository {
    private let client: APIClient

    init(client: APIClient = .shared) {
        self.client = client
    }

    func loadExpenseDetail(ref: TransactionRef) async throws -> APIClient.ExpenseDetail {
        guard ref.resourceType == .expense else {
            throw NSError(domain: "PersonalTransactionRepository", code: 400, userInfo: [NSLocalizedDescriptionKey: "Income detail uses activity payload."])
        }
        return try await client.getExpense(momentId: ref.momentId, expenseId: ref.resourceId)
    }

    func updateExpense(
        ref: TransactionRef,
        amount: String? = nil,
        currencyCode: String? = nil,
        description: String? = nil,
        merchantName: String? = nil,
        categoryCode: String? = nil,
        subcategoryCode: String? = nil,
        financialAccountId: String? = nil,
        paymentMethodCode: String? = nil,
        effectiveAt: String? = nil,
        recurringScheduleId: String? = nil
    ) async throws -> APIClient.UpdateExpenseResult {
        guard ref.resourceType == .expense else {
            throw NSError(domain: "PersonalTransactionRepository", code: 400, userInfo: [NSLocalizedDescriptionKey: "Use income endpoints for INCOME resources."])
        }
        return try await client.updateExpense(
            momentId: ref.momentId,
            expenseId: ref.resourceId,
            amount: amount,
            currencyCode: currencyCode,
            description: description,
            merchantName: merchantName,
            categoryCode: categoryCode,
            subcategoryCode: subcategoryCode,
            financialAccountId: financialAccountId,
            paymentMethodCode: paymentMethodCode,
            effectiveAt: effectiveAt,
            recurringScheduleId: recurringScheduleId
        )
    }

    func void(ref: TransactionRef) async throws {
        switch ref.resourceType {
        case .expense:
            _ = try await client.voidExpense(momentId: ref.momentId, expenseId: ref.resourceId)
        case .income:
            _ = try await client.voidPersonalIncome(momentId: ref.momentId, incomeId: ref.resourceId)
        default:
            throw NSError(domain: "PersonalTransactionRepository", code: 400, userInfo: [NSLocalizedDescriptionKey: "Unsupported resource type."])
        }
    }

    func attachMedia(ref: TransactionRef, bytes: Data, contentType: String = "image/jpeg") async throws -> APIClient.ExpenseAttachment {
        guard ref.resourceType == .expense else {
            throw NSError(domain: "PersonalTransactionRepository", code: 400, userInfo: [NSLocalizedDescriptionKey: "Attachments only supported on expenses."])
        }
        return try await client.uploadAndAttachExpenseMedia(
            momentId: ref.momentId,
            expenseId: ref.resourceId,
            bytes: bytes,
            contentType: contentType
        )
    }
}

File: apk/app/src/main/java/com/example/momentra/data/repository/PersonalTransactionRepository.kt
package com.example.momentra.data.repository

import com.example.momentra.data.api.CreateExpenseResultDto
import com.example.momentra.data.api.ExpenseAttachmentDto
import com.example.momentra.data.api.ExpenseDetailDto
import com.example.momentra.data.api.PersonalIncomeResultDto
import java.util.UUID

/** GET / PATCH / DELETE loader for personal expense + income transactions. */
class PersonalTransactionRepository(
    private val slice: PersonalSliceRepository = PersonalSliceRepository(),
) {
    suspend fun loadDetail(ref: TransactionRef): Result<ExpenseDetailDto> = when (ref.resourceType) {
        TransactionResourceType.EXPENSE -> slice.getExpense(ref.momentId, ref.resourceId)
        else -> Result.failure(IllegalArgumentException("Income detail load uses activity payload until GET income ships on client."))
    }

    suspend fun updateExpense(
        ref: TransactionRef,
        amount: String? = null,
        currencyCode: String? = null,
        description: String? = null,
        merchantName: String? = null,
        categoryCode: String? = null,
        subcategoryCode: String? = null,
        financialAccountId: String? = null,
        paymentMethodCode: String? = null,
        effectiveAt: String? = null,
        recurringScheduleId: String? = null,
    ): Result<CreateExpenseResultDto> {
        require(ref.resourceType == TransactionResourceType.EXPENSE) { "Use income endpoints for INCOME resources." }
        return slice.updateExpense(
            momentId = ref.momentId,
            expenseId = ref.resourceId,
            amount = amount,
            currencyCode = currencyCode,
            description = description,
            merchantName = merchantName,
            categoryCode = categoryCode,
            subcategoryCode = subcategoryCode,
            financialAccountId = financialAccountId,
            paymentMethodCode = paymentMethodCode,
            effectiveAt = effectiveAt,
            recurringScheduleId = recurringScheduleId,
        )
    }

    suspend fun voidExpense(ref: TransactionRef): Result<CreateExpenseResultDto> {
        require(ref.resourceType == TransactionResourceType.EXPENSE) { "Use voidIncome for INCOME resources." }
        return slice.voidExpense(ref.momentId, ref.resourceId)
    }

    suspend fun voidIncome(ref: TransactionRef): Result<PersonalIncomeResultDto> {
        require(ref.resourceType == TransactionResourceType.INCOME) { "Use voidExpense for EXPENSE resources." }
        return slice.voidPersonalIncome(ref.momentId, ref.resourceId)
    }

    suspend fun attachMedia(
        ref: TransactionRef,
        bytes: ByteArray,
        contentType: String = "image/jpeg",
        idempotencyKey: String = UUID.randomUUID().toString(),
    ): Result<ExpenseAttachmentDto> {
        require(ref.resourceType == TransactionResourceType.EXPENSE) { "Attachments only supported on expenses." }
        return slice.uploadAndAttachExpenseMedia(
            momentId = ref.momentId,
            expenseId = ref.resourceId,
            bytes = bytes,
            contentType = contentType,
            idempotencyKey = idempotencyKey,
        )
    }
}
